### ISSUE TITLE
Shortest path weight calculated by LM deviates from the one calculated by Dijkstra.

### ISSUE THREAD

**Issue by easbar** [graphhopper/graphhopper#1557]
Following the discussion in the [forum](https://discuss.graphhopper.com/t/alt-stopping-criterion-incomplete/4129/4), I tried running random queries ALT vs Dijkstra and found cases where there is a difference, see here: 1f34cd6307f01b3f0f04fa458f134dd061453243

However, there do not seem to be deviations between AStarBidirection and Dijkstra, so not sure if the problems are due to the stopping critirion which is the same for ALT and AStarBidirection.

**Comment by easbar** [comment]
I was wondering if the `buildRandomGraph` method has a problem, but modifying the distances should be ok, since they are only increased ? https://github.com/graphhopper/graphhopper/blob/022697da8dfc9ab6e6537fbc7b81efc55ac2e142/core/src/main/java/com/graphhopper/util/GHUtility.java#L189 Also the method uses `Helper.DIST_PLANE` to calculate the distances (should be ok because this is what the WeightApproximators use as well).

**Comment by karussell** [comment]
> so not sure if the problems are due to the stopping critirion which is the same for ALT and AStarBidirection

The suggestion in the forum was that we have to use a different stopping criterion for ALT, so your observation is an indication there is a problem with it for ALT

> but modifying the distances should be ok, since they are only increased ?

Yes, increasing the weight shouldn't be a problem for A*

> Also the method uses Helper.DIST_PLANE to calculate the distances (should be ok because this is what the WeightApproximators use as well).

Yes, should also be okayish for "smaller" distances.

**Comment by easbar** [comment]
> The suggestion in the forum was that we have to use a different stopping criterion for ALT, so your observation is an indication there is a problem with it for ALT

Are the active landmarks fixed per query or can they change during the query (onHeuristicChange) ? This could have implications on the stopping criterion. Then again I used `numLandmarks=numActiveLandmarks=4` in the tests, so there is nothing to change. 

The slides mentioned in the forum do not say there needs to be done anything particular for ALT, even though the stopping criterion is explained in the ALT chapter (but more generally for bidirectional AStar).

**Comment by karussell** [comment]
(renamed title from ALT to LM as the similarity to alternative routes is a bit ugly ;) Please revert if bad)

> Are the active landmarks fixed per query or can they change during the query (onHeuristicChange) ? 

They are fixed.

> The slides mentioned in the forum do not say there needs to be done anything particular for ALT

Ah, yes. Sorry, I remembered it incorrectly.

There is a problem with virtual nodes, see #1532

**Comment by karussell** [comment]
Another problem could be the precision or rounding issues that I did overlook when implementing this.

**Comment by easbar** [comment]
Ok I think we can rule out #1532 because the QueryGraph is not involved in the test.

**Comment by karussell** [comment]
Yes, at least for the "shortest" test you did, not necessarily for the problem discussed in the forum.

**Comment by easbar** [comment]
Thanks a lot for your explanation! Do you have a suggestion how to obtain the factor in `AStarBidirection#finished` ? I tried briefly with a fixed value, but I had to use a rather large number to make the tests I mentioned above pass. Did you manage to fix the problems mentioned in the forum with this approach ?

**Comment by aoles** [comment]
Thank you for getting back to me! 

In order to obtain the scaling factor I've added `public double LMApproximator.getFactor()` which I then call in `AStarBidirection.initTo` to set `private double AStarBidirection.approximatorOffset` like so:
```
if (weightApprox.getApproximation() instanceof LMApproximator)
  approximatorOffset = 2 * ((LMApproximator) weightApprox.getApproximation()).getFactor();
```

The corresponding stop condition reads
```
currFrom.weight + currTo.weight - approximatorOffset >= bestPath.getWeight();
```

I can confirm that this hack fixes all the issues I described in the forum.

**Comment by aoles** [comment]
The fix described above works only in the old approach where the from/to weights were stored independently. It seems that the new approach introduced in https://github.com/graphhopper/graphhopper/pull/1376 of storing only one of the two weights, say `from`, and only the difference to the other one `to` has a serious flaw in case the difference overflows the negative `DELTA_MIN`. This happens when `to < from` and `to - from < DELTA_MIN`. Then the reconstructed weight `to' = from + DELTA_MIN` will be higher than the original `to` value violating the landmark condition that the weights must never overestimate the distance. On the other hand, the overflow of the positive `DELTA_MAX` is not an issue because then the distance is just underestimated.

In fact, when I reserve more space for the delta storage by setting `FROM_WEIGHT_BITS = 16` all of the previously failing single-graph tests in [`CompareAlgoVsDijkstraTest`](https://github.com/graphhopper/graphhopper/blob/algos_vs_dijkstra/core/src/test/java/com/graphhopper/routing/CompareAlgoVsDijkstraTest.java) pass. However, the `randomGraph()` test still fails after a few iterations.

Unlike in the former approach of storing absolute weights where the precision loss could be compensated by an offset term in the routing algorithm stop condition, I don't see an easy way to ensure the correctness of the algorithm when a possibly capped difference is stored.

**Comment by sfendrich** [comment]
So the problem is that, if `DELTA_MIN`is underflowed, we store a value that is too large and the A* heuristics becomes incorrect, right?
One possible solution would be to remember (from the landmarks computation) the worst value (`WORST`) affected by the condition `to < from` and `to - from < DELTA_MIN`. Whenever `DELTA_MIN` appears inthe heuristics computation, it could be replaced by `WORST`, so the true distance is underestimated and the heuristics should be correct.

In addition, the discussion of #1376 mentions that a bit-combination of 16/16 worked better than 18/14. If so, I wonder what is the benefit of the 16/16 bit over the former solution of storing `from` and `to`in two shorts. Can we be certain that the performance gain of the newer solution is not influenced by A* stopping too early due to the precision loss?

**Comment by karussell** [comment]
Thanks @aoles for finding this issue. This is indeed tricky to fix. Maybe with a special value like we already have for "infinity".

> the discussion of #1376 mentions that a bit-combination of 16/16 worked better than 18/14
> I wonder what is the benefit of the 16/16 bit over the former solution of storing from and toin two shorts

@sfendrich see this comment @baumboi:

>  With `PRECISION = 16` we have the same accuracy and factor we had in the old system with two shorts, but less weightings maxed out (which means more precision on long distances).
> With `PRECISION = 18` we'd get better precision on short and normal distances, but have at least as many weights maxed out as before (probably even more because of the quite small size of the delta values).

So, if the delta solution would be better because of the problem @aoles mentioned, then we would get even better performance for 18/14 IMO.

**Comment by michaz** [comment]
Hey @aoles, thank you so much (again) for your analysis. I'm cautiously optimistic that we have this under control as of #1749.

(Turned out there is another reason why we need the "factor" in the stopping criterion. Even if the approximator would always round down, it would still violate the triangle inequality, with a similar result (we terminate early) and the same solution.)

**Comment by aoles** [comment]
You are most welcome @michaz , thanks a lot for the update!

Awesome work! Could you maybe briefly summarize what the other issue is and why there is still need for an offset in the stopping condition even with an always underestimating approximator?

Cheers,
Andrzej

**Comment by michaz** [comment]
> Could you maybe briefly summarize what the other issue is and why there is still need for an offset in the stopping condition even with an always underestimating approximator?

It's _approximately_ this: For the stopping criterion to work, it is sufficient that the approximator only underestimates. But only if, additionally, the top heap weight increases monotonically. This can fail to happen if the potential function between u and v decreases by a higher amount than the edge weight of (u,v). And this can happen at least when (u,v) is shorter than the "factor", no matter how we round.

In that case, the "corrected" edge weight (real edge weight minus difference of potential function) is negative, which means that the stopping criterion which just looks at the top heap weight doesn't work anymore. Same solution: Require the top heap weight to be higher by the maximum amount it can decrease. By "factor".

**Comment by michaz** [comment]
Fixed by #1749.
